community/kernel/src/main/java/org/neo4j/kernel/impl/api/TwoPhaseNodeForRelationshipLocking.java

```diff
@@ -82,24 +81,48 @@ void lockAllNodesAndConsumeRelationships( long nodeId, final KernelStatement sta
         while ( retry );
     }
 
-    private void lockAllNodes( KernelStatement state )
+    private void collectAndSortNodeIds( long nodeId, KernelStatement state ) throws EntityNotFoundException
+    {
+        PrimitiveLongSet nodeIdSet = Primitive.longSet();
+        nodeIdSet.add( nodeId );
+
+        try ( Cursor<NodeItem> node = ops.nodeCursorById( state, nodeId ) )
+        {
+            try ( Cursor<RelationshipItem> rels = ops.nodeGetRelationships( state, node.get(), Direction.BOTH ) )
+            {
+                while ( rels.next() )
+                {
+                    RelationshipItem rel = rels.get();
+                    if ( firstRelId == NO_SUCH_RELATIONSHIP )
+                    {
+                        firstRelId = rel.id();
+                    }
+
+                    nodeIdSet.add( rel.startNode() );
+                    nodeIdSet.add( rel.endNode() );
+                }
+            }
+        }
+
+        long[] nodeIds = PrimitiveArrays.of( nodeIdSet );
+        Arrays.sort( nodeIds );
+        this.sortedNodeIds = nodeIds;
+    }
+
+    private void lockAllNodes( KernelStatement state, long[] nodeIds )
     {
-        PrimitiveLongIterator nodeIdIterator = nodeIds.iterator();
-        while ( nodeIdIterator.hasNext() )
+        for ( long nodeId : nodeIds )
         {
-            state.locks().optimistic()
-                    .acquireExclusive( state.lockTracer(), ResourceTypes.NODE, nodeIdIterator.next() );
+            state.locks().optimistic().acquireExclusive( state.lockTracer(), ResourceTypes.NODE, nodeId );
         }
     }
 
-    private void unlockAllNodes( KernelStatement state )
+    private void unlockAllNodes( KernelStatement state, long[] nodeIds  )
     {
-        PrimitiveLongIterator iterator = nodeIds.iterator();
-        while ( iterator.hasNext() )
+        for ( long nodeId : nodeIds )
         {
-            state.locks().optimistic().releaseExclusive( ResourceTypes.NODE, iterator.next() );
+            state.locks().optimistic().releaseExclusive( ResourceTypes.NODE, nodeId );
```

Comment:
You could use one call since the id argument is a vararg

Reply:
Actually, for some reason `releaseExclusive` is not varargs. Maybe that should be changed, but because it has 7 implementations I would rather not do it in this PR.